src/services/utils.ts

```diff
@@ -37,23 +38,64 @@ type DependencyDefinition = {
   arguments?: Array<RequiredToken | OptionalToken>
   shared?: boolean
   labels?: Token[]
+  decorates?: Token
 }
 export type ServiceDefinition = [Token, DependencyDefinition]
+export type ServiceConfigurator<T = Record<string, Token>> = ($: T) => ServiceDefinition[]
 export type Alias<T extends (...args: never[]) => unknown> = (...rest: Parameters<T>) => ReturnType<T>
-
+export type Decorator<T extends TokenValue> = TokenType<T>
+export type ReturnDecorated<T extends TokenValue> = () => TokenType<T>
 export const container = createContainer()
 
 //
 export const merge = (...definitions: Array<ServiceDefinition[]>): ServiceDefinition[] => {
   return [...new Map([...definitions.flat()]).entries()]
 }
-export const build = (...entries: Array<ServiceDefinition[]>): ServiceDefinition[] => {
-  const merged = merge(...entries)
+const decorate = (entries: ServiceDefinition[]) => {
+  const map = new Map(entries)
+  entries.forEach(([_token, definition]) => {
+    if (typeof definition.decorates !== 'undefined') {
+      const decoratedToken = definition.decorates
+      const decoratedDefinition = map.get(decoratedToken)
+      if (typeof decoratedDefinition !== 'undefined') {
+        // set up a structure that allows us to only get the decorated service
+        // when we want to. i.e. we might want to call/get the service then execute more code, or
+        // we might want to call/get the service last after executing code
+        const { labels, ...decorated } = decoratedDefinition
+        type DecoratedToken = typeof decoratedToken
+        const TOKEN = token<TokenType<DecoratedToken>>('inner')
+        const WRAPPER_TOKEN = token<ReturnDecorated<DecoratedToken>>('inner.wrapper')
+        map.set(TOKEN, {
+          ...decorated,
+        })
+        map.set(WRAPPER_TOKEN, {
+          service: (): ReturnDecorated<DecoratedToken> => () => get(TOKEN),
+        })
+
+        map.set(decoratedToken,
+          {
+            ...definition,
+            arguments: [
+              WRAPPER_TOKEN,
+              // adding further arguments is currently unsupported
+              // ...(definition.arguments || []),
```

Comment:
We don't need to do this right now, but if we do need this (decorators with arguments), this is where it would be done.